### PR TITLE
Add Random

### PR DESCRIPTION
**Pull request by JonasWanke**

### CODE REVIEW

File: packages/core/src/random/source.candy [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  fun generateIntegers(length: Int): List<Int>
+
+  // TODO(JonasWanke): use bytes when we have them
+  // fun generateBytes(length: Int): List<Int>
+}
+
+builtin class DefaultRandomSource {
+  static fun withSeed(seed: Int): This
+}
+builtin impl DefaultRandomSource: RandomSource
```
Comment (MarcelGarus) [inline packages/core/src/random/source.candy]: Do we really want to expose `DefaultRandomSource` to the rest of the Candy ecosystem, or should we rather provide static methods on `RandomSource`, as in `RandomSource.default()` and `RandomSource.defaultWithSeed(...)`?

File: packages/core/src/random/source.candy [hunk 2]
```diff
@@ -0,0 +1,14 @@
+use ...collections
+use ...primitives
+
+trait RandomSource {
+  fun generateIntegers(length: Int): List<Int>
+
+  // TODO(JonasWanke): use bytes when we have them
+  // fun generateBytes(length: Int): List<Int>
```
Comment (MarcelGarus) [inline packages/core/src/random/source.candy]: I'm not sure about this one. I think we should offer randomness at smaller granularities:
- `generateBool` for generating a bit
- `generateByte` for generating a byte
- `generateBytes` for generating an arbitrary number of bytes

Of course, for now, the byte generators could return `int`s up to 255 instead.

Some advantages:

- We could provide default implementations for everything except `generateBool`. That makes it very easy to implement `RandomSource` while keeping the option to write more efficient `RandomSource`s. ("Very easy" as in "just write a method that returns `true` 50% of the time".)
- Providing randomness on bit granularity also has the advantage that we are more careful about using up entropy. When we implement a secure `RandomSource`, we'll probably use something like /dev/random on Linux and we could use bitshifting to give out single bits of entropy. That seems especially useful for generating bools and enum variants without consuming a whole byte of entropy every time.

That being said, I'm also okay with `generateByte` being the one without a default implementation. Most implementations work on byte-granularity anyways and for those that are extra careful about using entropy, they can implement `generateBool` too. For example, our default number generator probably doesn't need a special `generateBool`, but the secure one does.

File: packages/compiler/lib/src/compilation/hir/type.dart [hunk 3]
```diff
@@ -125,6 +126,10 @@ abstract class CandyType with _$CandyType {
   static const implies =
       UserCandyType(ModuleId.coreOperatorsLogical, 'Implies');
 
+  // reflection
```
Comment (MarcelGarus) [inline packages/compiler/lib/src/compilation/hir/type.dart]:
```suggestion
  // random
```

**Review by MarcelGarus — CHANGES_REQUESTED** [review]

File: packages/core/src/random/source.candy (hunk 1 above)
Comment (JonasWanke) [reply packages/core/src/random/source.candy]: It's private already

File: packages/core/src/random/source.candy (hunk 2 above)
Comment (JonasWanke) [reply packages/core/src/random/source.candy]: It's now changed to provide the three methods mentioned above, with `generateByte` being the default used by the other two.

**Review by MarcelGarus — APPROVED** [review]

**Pull request merged**